### PR TITLE
Split SDK code files, catch up to rust SDK v0.2.2, other fixes

### PR DESCRIPTION
**Pull request by quwenjie**
Split `sdk_a.py` to multiple source files.

Follow the Rust SDK version from v0.2.0 to v0.2.2:
- Add version to RequestInfo: https://github.com/CoLearn-Dev/colink-sdk-rust-dev/pull/29
- Instant server: https://github.com/CoLearn-Dev/colink-sdk-rust-dev/pull/26
- Protocol Attach: https://github.com/CoLearn-Dev/colink-sdk-rust-dev/pull/26
- Update CI workflow:  https://github.com/CoLearn-Dev/colink-sdk-rust-dev/pull/26

Fix update registries: https://github.com/CoLearn-Dev/colink-sdk-rust-dev/pull/30

Fix use grpcio-tools strange build bug v1.48.0 triggered by CI: upgrade to v1.50.0

### CODE REVIEW

File: colink/switch_to_generated_user.py [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+    pk, sk = generate_user()
+    core_pub_key = self.request_info().core_public_key
+    signature_timestamp, sig = prepare_import_user_signature(
+        pk, sk, core_pub_key, expiration_timestamp
+    )
+    return self.import_user(pk, signature_timestamp, expiration_timestamp, sig)
+
+
+def switch_to_generated_user(self):
+    self.__init__(coreaddr=self.core_addr, jwt=self.generate_user_and_import())
```
Comment (camelop) [inline colink/switch_to_generated_user.py:21]: just don't... don't do this...

File: colink/wait_user_init_func.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+    queue_name = self.subscribe(is_initialized_key, start_timestamp)
+
+    subscriber = self.new_subscriber(queue_name)
+    while True:
+        data = subscriber.get_next()
+        logging.info(f"Received [{byte_to_str(data)}]")
+        message = CL.SubscriptionMessage.FromString(data)
+        if message.change_type != "delete" and message.payload[0] == 1:
+            break
+        self.unsubscribe(queue_name)
```
Comment (camelop) [inline colink/wait_user_init_func.py:24]: ?

File: colink/sdk_p.py [hunk 3]
```diff
@@ -2,15 +2,21 @@
...
-def thread_func(protocol_and_role, cl, user_func):
-    cl_app = CoLinkProtocol(protocol_and_role, cl, user_func)
-    cl_app.start()
+def thread_func(q, protocol_and_role, cl, user_func):
+    try:
+        cl_app = CoLinkProtocol(protocol_and_role, cl, user_func)
+        cl_app.start()
+    except Exception as e:
+        q.put(e)
+        return
```
Comment (camelop) [inline colink/sdk_p.py:19]: ?

File: colink/sdk_p.py [hunk 4]
```diff
@@ -58,28 +65,31 @@ def run(self):
...
-                raise e
+            t.start()
+        while True:
+            if q.empty():
+                if threading.active_count() == 1:
+                    break
+            else:
+                err = q.get()
+                raise err
         return
```
Comment (camelop) [inline colink/sdk_p.py:87]: ?

File: colink/instant_server.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+
+        host_token = open(os.path.join(working_dir, "host_token.txt"), "r").read()
+        return InstantServer(
+            id=instant_server_id,
+            port=port,
+            host_token=host_token,
+            process=child,
+        )
+
+    def __del(self):
```
Comment (camelop) [inline colink/instant_server.py:78]: rename?

**Review by camelop — CHANGES_REQUESTED** [review]
Looks great! We finally refactored the structure 🎉 
Thanks and please see the detailed comment.

File: colink/sdk_a.py [hunk 6]
```diff
@@ -38,695 +46,379 @@ def get_next(self) -> bytes:
...
+    return self.run_task_with_expiration_time(
+        protocol_name,
+        protocol_param,
+        participants,
+        require_agreement,
+        get_time_stamp() + 86400,
+    )
+
+
+def run_task_with_expiration_time(
```
[the remark this answers is not in the document]
Comment (quwenjie) [reply colink/sdk_a.py:236]: resolved

File: colink/sdk_a.py [hunk 7]
```diff
@@ -38,695 +46,379 @@ def get_next(self) -> bytes:
...
+    )
+
+
+def new_subscriber(self, queue_name: str) -> CoLinkSubscriber:
+    mq_uri = self.request_info().mq_uri
+    subscriber = CoLinkSubscriber(mq_uri, queue_name)
+    return subscriber
+
+
+def ca_certificate(self, ca_certificate: str):
```
[the remark this answers is not in the document]
Comment (quwenjie) [reply colink/sdk_a.py:320]: Have moved.

File: examples/user_exchange_jwt.py [hunk 8]
```diff
@@ -1,6 +1,6 @@
 import sys
+from colink import CoLink
```
[the remark this answers is not in the document]
Comment (quwenjie) [reply examples/user_exchange_jwt.py:2]: Resolved.

File: colink/lock_key.py [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+    return (key, rnd_num)
+
+
+def unlock(self, lock_token: Tuple[str, int]):
+    key, rnd_num = lock_token
+    rnd_num_in_storage = byte_to_int(self.read_entry("_lock:{}".format(key)))
+    if rnd_num_in_storage == rnd_num:
+        self.delete_entry("_lock:{}".format(key))
+    else:
+        logging.error("Invalid token.")
```
[the remark this answers is not in the document]
Comment (quwenjie) [reply colink/lock_key.py:42]: Resolved.

File: colink/switch_to_generated_user.py (hunk 1 above)
Comment (quwenjie) [reply colink/switch_to_generated_user.py:21]: Resolved

**Comment by quwenjie** [comment]
I also tested the delete functionality of instant server and it worked well.

File: colink/participant_id.py [hunk 10]
```diff
@@ -0,0 +1,9 @@
+import colink as CL
+from typing import List
+
+
+def get_participant_id(self, participants: List[CL.Participant]) -> int:
```
[the remark this answers is not in the document]
Comment (quwenjie) [reply colink/participant_id.py:5]: Have synced.

File: colink/wait_user_init_func.py (hunk 2 above)
Comment (quwenjie) [reply colink/wait_user_init_func.py:24]: Resolved.

File: colink/sdk_p.py (hunk 3 above)
Comment (quwenjie) [reply colink/sdk_p.py:19]: Resolved.

File: colink/sdk_p.py (hunk 4 above)
Comment (quwenjie) [reply colink/sdk_p.py:87]: Resolved.

File: colink/instant_server.py (hunk 5 above)
Comment (quwenjie) [reply colink/instant_server.py:78]: Resolved.

File: colink/__init__.py [hunk 11]
```diff
@@ -1,13 +1,13 @@
 from typing import Tuple
 from colink.colink_pb2 import *
-from colink.colink_pb2_grpc import *
+from colink.colink_pb2_grpc import CoLinkStub, CoLinkServicer
```
Comment (quwenjie) [inline colink/__init__.py:3]: In colink_pb2_grpc there is also a CoLink class, to resolve conflict we can't import *.

File: colink/__init__.py [hunk 12]
```diff
@@ -78,3 +91,7 @@ def __init__(
         switch_to_generated_user,
     )
     from .wait_user_init_func import wait_user_init
+
+
+from .protocol import ProtocolOperator
```
Comment (camelop) [inline colink/__init__.py:96]: Can we put this on top? It's unnatural to have import at the bottom of a file.

**Review by camelop — CHANGES_REQUESTED** [review]
Most things are wonderful! Once you move that import, I think we're ready to merge.
Next, we can take a look at the automatic serialization, as well as other Rust updates.
Really appreciate the effort!

**Review by camelop — APPROVED** [review]
Looks great!

**Pull request merged**